# Data preprocessing

In [1]:
import numpy as np
import pandas as pd

In [10]:
df = pd.read_csv("data/1_Poland_Lithuania_Mixed_finished.csv")

In [6]:
df2 = pd.read_csv("data/2_Italy_France_Mixed.csv")
df2

Observation id         Observation date  \
0    2_Italy_France_Mixed  2025-05-19 08:35:09.703   
1    2_Italy_France_Mixed  2025-05-19 08:35:09.703   
2    2_Italy_France_Mixed  2025-05-19 08:35:09.703   
3    2_Italy_France_Mixed  2025-05-19 08:35:09.703   
4    2_Italy_France_Mixed  2025-05-19 08:35:09.703   
..                    ...                      ...   
529  2_Italy_France_Mixed  2025-05-19 08:35:09.703   
530  2_Italy_France_Mixed  2025-05-19 08:35:09.703   
531  2_Italy_France_Mixed  2025-05-19 08:35:09.703   
532  2_Italy_France_Mixed  2025-05-19 08:35:09.703   
533  2_Italy_France_Mixed  2025-05-19 08:35:09.703   

                               Description  Observation duration  \
0    Team A is France and team B is Italy.                5389.8   
1    Team A is France and team B is Italy.                5389.8   
2    Team A is France and team B is Italy.                5389.8   
3    Team A is France and team B is Italy.                5389.8   
4    Team A is France and team B is Italy.                5389.8   
..                                     ...                   ...   
529  Team A is France and team B is Italy.                5389.8   
530  Team A is France and team B is Italy.                5389.8   
531  Team A is France and team B is Italy.                5389.8   
532  Team A is France and team B is Italy.                5389.8   
533  Team A is France and team B is Italy.                5389.8   

    Observation type                                             Source  \
0      Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
1      Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
2      Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
3      Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
4      Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
..               ...                                                ...   
529    Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
530    Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
531    Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
532    Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   
533    Media file(s)  player #1:C:/Users/basti/Desktop/UF_Thesis/2_I...   

     Time offset (s)  Media duration (s)   FPS  Subs_A  ...  Subject  \
0                0.0              7220.0  30.0      16  ...      NaN   
1                0.0              7220.0  30.0      16  ...      NaN   
2                0.0              7220.0  30.0      16  ...      NaN   
3                0.0              7220.0  30.0      16  ...      NaN   
4                0.0              7220.0  30.0      16  ...      NaN   
..               ...                 ...   ...     ...  ...      ...   
529              0.0              7220.0  30.0      16  ...      NaN   
530              0.0              7220.0  30.0      16  ...      NaN   
531              0.0              7220.0  30.0      16  ...      NaN   
532              0.0              7220.0  30.0      16  ...      NaN   
533              0.0              7220.0  30.0      16  ...      NaN   

                  Behavior  Behavioral category Modifier #1  Behavior type  \
0               First half                  NaN         NaN          START   
1    Time_on_offense_teamA                  NaN         NaN          START   
2              Pull_pickup                  NaN      RZ_Def          POINT   
3          Successful_pass                  NaN      RZ_Def          POINT   
4          Successful_pass                  NaN     Mid_Def          POINT   
..                     ...                  ...         ...            ...   
529        Successful_pass                  NaN     Mid_Off          POINT   
530        Successful_pass                  NaN      RZ_Off          POINT   
531                  Score                  NaN         NaN          POINT   

In [11]:
df

Observation id         Observation date  \
0    1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
1    1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
2    1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
3    1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
4    1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
..                        ...                      ...   
525  1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
526  1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
527  1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
528  1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   
529  1_Poland_Lithuania_Mixed  2025-05-02 16:07:56.825   

                                           Description  Observation duration  \
0    The game of Poland against Lithuania in the mi...              5990.167   
1    The game of Poland against Lithuania in the mi...              5990.167   
2    The game of Poland against Lithuania in the mi...              5990.167   
3    The game of Poland against Lithuania in the mi...              5990.167   
4    The game of Poland against Lithuania in the mi...              5990.167   
..                                                 ...                   ...   
525  The game of Poland against Lithuania in the mi...              5990.167   
526  The game of Poland against Lithuania in the mi...              5990.167   
527  The game of Poland against Lithuania in the mi...              5990.167   
528  The game of Poland against Lithuania in the mi...              5990.167   
529  The game of Poland against Lithuania in the mi...              5990.167   

    Observation type                                             Source  \
0      Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
1      Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
2      Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
3      Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
4      Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
..               ...                                                ...   
525    Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
526    Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
527    Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
528    Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   
529    Media file(s)  player #1:C:/Users/basti/Desktop/Emmie/Poland ...   

     Time offset (s)  Media duration (s)   FPS  Subs_A  ...  Subject  \
0                0.0              6485.0  30.0      17  ...      NaN   
1                0.0              6485.0  30.0      17  ...      NaN   
2                0.0              6485.0  30.0      17  ...      NaN   
3                0.0              6485.0  30.0      17  ...      NaN   
4                0.0              6485.0  30.0      17  ...      NaN   
..               ...                 ...   ...     ...  ...      ...   
525              0.0              6485.0  30.0      17  ...      NaN   
526              0.0              6485.0  30.0      17  ...      NaN   
527              0.0              6485.0  30.0      17  ...      NaN   
528              0.0              6485.0  30.0      17  ...      NaN   
529              0.0              6485.0  30.0      17  ...      NaN   

                  Behavior  Behavioral category Modifier #1  Behavior type  \
0               First half                  NaN         NaN          START   
1    Time_on_offense_teamA                  NaN         NaN          START   
2          Successful_pass                  NaN         Mid          POINT   
3          Successful_pass                  NaN         Mid          POINT   
4          Successful_pass                  NaN         Mid          POINT   
..                     ...                  ...         ...            ...   
525        Successful_pass                  N

In [12]:
df = df.drop(columns=['Observation date', 'Observation duration', 'Description', 'Observation type', 'Source', 'Time offset (s)', 'Media duration (s)', 'FPS', 'Subject', 'Behavioral category', 'Media file name', 'Image index', 'Image file path'])

df = df.rename(columns={"Behavior" : "Behaviour", "Modifier #1" : "Field_Loc", "Behavior type" : "B_Type"})

df

Observation id  Subs_A  Early win  Subs_B  \
0    1_Poland_Lithuania_Mixed      17          1      16   
1    1_Poland_Lithuania_Mixed      17          1      16   
2    1_Poland_Lithuania_Mixed      17          1      16   
3    1_Poland_Lithuania_Mixed      17          1      16   
4    1_Poland_Lithuania_Mixed      17          1      16   
..                        ...     ...        ...     ...   
525  1_Poland_Lithuania_Mixed      17          1      16   
526  1_Poland_Lithuania_Mixed      17          1      16   
527  1_Poland_Lithuania_Mixed      17          1      16   
528  1_Poland_Lithuania_Mixed      17          1      16   
529  1_Poland_Lithuania_Mixed      17          1      16   

                 Behaviour Field_Loc B_Type      Time Comment  
0               First half       NaN  START   164.800     NaN  
1    Time_on_offense_teamA       NaN  START   295.667     NaN  
2          Successful_pass       Mid  POINT   297.467     NaN  
3          Successful_pass       Mid  POINT   301.267     NaN  
4          Successful_pass       Mid  POINT   306.133     NaN  
..                     ...       ...    ...       ...     ...  
525        Successful_pass    RZ_Off  POINT  6143.333     NaN  
526        Successful_pass    RZ_Off  POINT  6148.067     NaN  
527                  Score       NaN  POINT  6153.067     NaN  
528            Second half       NaN   STOP  6154.967     NaN  
529  Time_on_offense_teamA       NaN   STOP  6154.967     NaN  

[530 rows x 9 columns]

In [13]:
df.to_csv("data/cleaned_df.csv", index = False)

Some randoms stats

In [14]:
df.loc[df['Behaviour'] == 'Score'].count()

Observation id    29
Subs_A            29
Early win         29
Subs_B            29
Behaviour         29
Field_Loc          0
B_Type            29
Time              29
Comment            0
dtype: int64

Maybe I need to make a table for team A and a table for team B.

In [8]:
# team_A = df.copy()

In [57]:
# Filter only Time_on_offense_teamA state rows
a_offense_rows = df[df['Behavior'] == 'Time_on_offense_teamA']

# Get list of start/stop timestamps in order
starts = a_offense_rows[a_offense_rows['Behavior type'] == 'START']['Time'].tolist()
stops  = a_offense_rows[a_offense_rows['Behavior type'] == 'STOP']['Time'].tolist()

# Zip into intervals
a_intervals = list(zip(starts, stops))

In [10]:
from datetime import timedelta

def subtract_intervals(base_intervals, sub_intervals):
    result = []
    for start, stop in base_intervals:
        temp = [(start, stop)]
        for sub_start, sub_stop in sub_intervals:
            new_temp = []
            for s, e in temp:
                # Skip non-overlapping
                if sub_stop <= s or sub_start >= e:
                    new_temp.append((s, e))
                else:
                    # Chop out the timeout
                    if sub_start > s:
                        new_temp.append((s, sub_start))
                    if sub_stop < e:
                        new_temp.append((sub_stop, e))
            temp = new_temp
        result.extend(temp)
    return result

# Get timeout intervals
timeout_rows = df[df['Behavior'] == 'Timeout']
timeout_starts = timeout_rows[timeout_rows['Behavior type'] == 'START']['Time'].tolist()
timeout_stops  = timeout_rows[timeout_rows['Behavior type'] == 'STOP']['Time'].tolist()
timeout_intervals = list(zip(timeout_starts, timeout_stops))

# Subtract timeouts from Team A offense
a_intervals_clean = subtract_intervals(a_intervals, timeout_intervals)


In [11]:
# Assuming continuous play (i.e., always one team on offense),
# get full game range from min to max timestamp
game_start = df['Time'].min()
game_end = df['Time'].max()

# Subtract a_intervals_clean from (game_start, game_end)
def invert_intervals(intervals, start, end):
    result = []
    prev_end = start
    for s, e in sorted(intervals):
        if s > prev_end:
            result.append((prev_end, s))
        prev_end = max(prev_end, e)
    if prev_end < end:
        result.append((prev_end, end))
    return result

b_intervals_clean = invert_intervals(a_intervals_clean, game_start, game_end)


In [12]:
def rows_in_intervals(df, intervals):
    return df[df['Time'].apply(lambda t: any(start <= t <= end for start, end in intervals))]

df_teamA_offense = rows_in_intervals(df, a_intervals_clean)
df_teamB_offense = rows_in_intervals(df, b_intervals_clean)

In [13]:
df_teamA_offense

Observation id  Subs_A  Early win  Subs_B  \
1    1_Poland_Lithuania_Mixed      17          1      16   
2    1_Poland_Lithuania_Mixed      17          1      16   
3    1_Poland_Lithuania_Mixed      17          1      16   
4    1_Poland_Lithuania_Mixed      17          1      16   
5    1_Poland_Lithuania_Mixed      17          1      16   
..                        ...     ...        ...     ...   
525  1_Poland_Lithuania_Mixed      17          1      16   
526  1_Poland_Lithuania_Mixed      17          1      16   
527  1_Poland_Lithuania_Mixed      17          1      16   
528  1_Poland_Lithuania_Mixed      17          1      16   
529  1_Poland_Lithuania_Mixed      17          1      16   

                  Behavior Modifier #1 Behavior type      Time Comment  
1    Time_on_offense_teamA         NaN         START   295.667     NaN  
2          Successful_pass         Mid         POINT   297.467     NaN  
3          Successful_pass         Mid         POINT   301.267     NaN  
4          Successful_pass         Mid         POINT   306.133     NaN  
5          Successful_pass         Mid         POINT   307.467     NaN  
..                     ...         ...           ...       ...     ...  
525        Successful_pass      RZ_Off         POINT  6143.333     NaN  
526        Successful_pass      RZ_Off         POINT  6148.067     NaN  
527                  Score         NaN         POINT  6153.067     NaN  
528            Second half         NaN          STOP  6154.967     NaN  
529  Time_on_offense_teamA         NaN          STOP  6154.967     NaN  

[423 rows x 9 columns]

In [14]:
df_teamA_offense.loc[df_teamA_offense['Behavior'] == 'Score'].count()

Observation id    15
Subs_A            15
Early win         15
Subs_B            15
Behavior          15
Modifier #1        0
Behavior type     15
Time              15
Comment            0
dtype: int64

In [15]:
df_teamB_offense.loc[df_teamB_offense['Behavior'] == 'Score'].count()

Observation id    15
Subs_A            15
Early win         15
Subs_B            15
Behavior          15
Modifier #1        0
Behavior type     15
Time              15
Comment            0
dtype: int64

In [16]:
df_teamB_offense

Observation id  Subs_A  Early win  Subs_B  \
0    1_Poland_Lithuania_Mixed      17          1      16   
1    1_Poland_Lithuania_Mixed      17          1      16   
7    1_Poland_Lithuania_Mixed      17          1      16   
8    1_Poland_Lithuania_Mixed      17          1      16   
9    1_Poland_Lithuania_Mixed      17          1      16   
..                        ...     ...        ...     ...   
489  1_Poland_Lithuania_Mixed      17          1      16   
490  1_Poland_Lithuania_Mixed      17          1      16   
491  1_Poland_Lithuania_Mixed      17          1      16   
492  1_Poland_Lithuania_Mixed      17          1      16   
493  1_Poland_Lithuania_Mixed      17          1      16   

                  Behavior Modifier #1 Behavior type      Time Comment  
0               First half         NaN         START   164.800     NaN  
1    Time_on_offense_teamA         NaN         START   295.667     NaN  
7          Forced_turnover         Mid         POINT   314.500     NaN  
8                     Pass         Mid         POINT   314.500     NaN  
9    Time_on_offense_teamA         NaN          STOP   314.500     NaN  
..                     ...         ...           ...       ...     ...  
489        Successful_pass         Mid         POINT  5715.933     NaN  
490        Successful_pass         Mid         POINT  5719.800     NaN  
491        Successful_pass         Mid         POINT  5769.167     NaN  
492                  Score         NaN         POINT  5773.700     NaN  
493  Time_on_offense_teamA         NaN         START  5943.367     NaN  

[210 rows x 9 columns]

Total game time (includes the half time break and timeouts).

Time of each half.

Time is in seconds.

Time of each team on offense.

In [17]:
game_start = df['Time'].min()
game_end = df['Time'].max()

total_gametime = (game_end - game_start) / 60

print("Total game time is:", total_gametime, "minutes.")

Total game time is: 99.83611666666665 minutes.


In [18]:
first_half_start = df[(df['Behavior'] == 'First half') & (df['Behavior type'] == 'START')]['Time'].iloc[0]
first_half_stop = df[(df['Behavior'] == 'First half') & (df['Behavior type'] == 'STOP')]['Time'].iloc[0]

first_half_dur = (first_half_stop - first_half_start) / 60
print("Total duration of the first half is:", first_half_dur, "minutes.")

first_half_df = df[(df['Time'] >= first_half_start) & (df['Time'] <= first_half_stop)]

Total duration of the first half is: 36.67278333333333 minutes.


In [19]:
second_half_start = df[(df['Behavior'] == 'Second half') & (df['Behavior type'] == 'START')]['Time'].iloc[0]
second_half_stop = df[(df['Behavior'] == 'Second half') & (df['Behavior type'] == 'STOP')]['Time'].iloc[0]

second_half_dur = (second_half_stop - second_half_start) / 60
print("Total duration of the second half is:", second_half_dur, "minutes.")

second_half_df = df[(df['Time'] >= second_half_start) & (df['Time'] <= second_half_stop)]

Total duration of the second half is: 57.8139 minutes.


In [40]:
total_gametime_excl_breaks = first_half_dur + second_half_dur

print(f"Total duration of the whole game (excluding breaks) is: {total_gametime_excl_breaks} minutes.")

Total duration of the whole game (excluding breaks) is: 94.48668333333333 minutes.


In [20]:
teamA_t_offense = df[(df['Behavior'] == 'Time_on_offense_teamA') & (df['Behavior type'] == 'START')]['Time'].iloc[0]
second_half_stop = df[(df['Behavior'] == 'Second half') & (df['Behavior type'] == 'STOP')]['Time'].iloc[0]

In [21]:
# Filter only rows related to Team A offense
a_offense = df[df['Behavior'] == 'Time_on_offense_teamA']

# Get start and stop times as lists
starts = a_offense[a_offense['Behavior type'] == 'START']['Time'].tolist()
stops  = a_offense[a_offense['Behavior type'] == 'STOP']['Time'].tolist()

# Safety check: they should match in length
if len(starts) != len(stops):
    raise ValueError("Mismatched number of START and STOP events for Team A offense.")

# Zip into intervals and sum durations
intervals = list(zip(starts, stops))
total_seconds = sum(stop - start for start, stop in intervals)

# Convert to minutes if needed
total_minutes = total_seconds / 60

print("Team A was on offense for:", total_minutes, "minutes.")


Team A was on offense for: 56.84223333333331 minutes.


In [22]:
# --- Step 1: Get Team A offense intervals ---
a_offense = df[df['Behavior'] == 'Time_on_offense_teamA']
starts = a_offense[a_offense['Behavior type'] == 'START']['Time'].tolist()
stops  = a_offense[a_offense['Behavior type'] == 'STOP']['Time'].tolist()

if len(starts) != len(stops):
    raise ValueError("Mismatched START and STOP for Team A offense.")

a_intervals = list(zip(starts, stops))

# --- Step 2: Get Timeout intervals ---
timeouts = df[df['Behavior'] == 'Timeout']
to_starts = timeouts[timeouts['Behavior type'] == 'START']['Time'].tolist()
to_stops  = timeouts[timeouts['Behavior type'] == 'STOP']['Time'].tolist()

if len(to_starts) != len(to_stops):
    raise ValueError("Mismatched START and STOP for timeouts.")

timeout_intervals = list(zip(to_starts, to_stops))


In [23]:
def subtract_intervals(base_intervals, subtracting_intervals):
    """Returns a list of intervals after removing overlaps with subtracting_intervals."""
    result = []
    for base_start, base_stop in base_intervals:
        temp = [(base_start, base_stop)]
        for sub_start, sub_stop in subtracting_intervals:
            new_temp = []
            for s, e in temp:
                if sub_stop <= s or sub_start >= e:
                    new_temp.append((s, e))  # No overlap
                else:
                    if sub_start > s:
                        new_temp.append((s, sub_start))
                    if sub_stop < e:
                        new_temp.append((sub_stop, e))
            temp = new_temp
        result.extend(temp)
    return result

# Get cleaned (timeout-free) intervals
a_clean_intervals = subtract_intervals(a_intervals, timeout_intervals)


In [42]:
total_a_seconds = sum(stop - start for start, stop in a_clean_intervals)
total_a_minutes = total_a_seconds / 60

print(f"Team A offense time (excluding timeouts): {total_a_minutes} minutes.")

Team A offense time (excluding timeouts): 50.167233333333314 minutes.


In [25]:
# Assume you already have a list of Team A offense intervals
a_offense = df[df['Behavior'] == 'Time_on_offense_teamA']
a_starts = a_offense[a_offense['Behavior type'] == 'START']['Time'].tolist()
a_stops  = a_offense[a_offense['Behavior type'] == 'STOP']['Time'].tolist()

if len(a_starts) != len(a_stops):
    raise ValueError("Mismatched START/STOP for Team A offense.")

a_intervals = list(zip(a_starts, a_stops))
a_intervals.sort()

# Step 1: Calculate full game range
game_start = df['Time'].min()
game_end = df['Time'].max()

# Step 2: Invert Team A intervals to get Team B's offense intervals
def invert_intervals(intervals, start, end):
    """Given a list of intervals, return the gaps between them within (start, end)."""
    result = []
    prev_end = start
    for s, e in sorted(intervals):
        if s > prev_end:
            result.append((prev_end, s))
        prev_end = max(prev_end, e)
    if prev_end < end:
        result.append((prev_end, end))
    return result

b_intervals = invert_intervals(a_intervals, game_start, game_end)


In [26]:
# Timeout intervals
timeouts = df[df['Behavior'] == 'Timeout']
to_starts = timeouts[timeouts['Behavior type'] == 'START']['Time'].tolist()
to_stops  = timeouts[timeouts['Behavior type'] == 'STOP']['Time'].tolist()

if len(to_starts) != len(to_stops):
    raise ValueError("Mismatched START/STOP for Timeout.")

timeout_intervals = list(zip(to_starts, to_stops))

# Reuse subtract_intervals function from earlier
b_clean_intervals = subtract_intervals(b_intervals, timeout_intervals)


In [27]:
total_b_seconds = sum(stop - start for start, stop in b_clean_intervals)
total_b_minutes = total_b_seconds / 60

print("Team B offense time (excluding timeouts):", total_b_minutes, "minutes.")


Team B offense time (excluding timeouts): 39.33776666666669 minutes.


In [ ]:
total_game_time = game_end - game_start
total_timeouts  = sum(to - from_ for from_, to in timeout_intervals)
total_offense_time = total_b_seconds + total_a_minutes  # From earlier for Team A

assert abs(total_game_time - total_timeouts - total_offense_time) < 1e-6, "Something's off"


In [29]:
# 7/5 10:06
# getting the indices of the rows where 

row_index = df.index[df['Behavior'] == 'Time_on_offense_teamA'].tolist()
print(row_index)

[1, 9, 16, 25, 29, 37, 40, 47, 59, 65, 72, 81, 87, 93, 96, 103, 105, 107, 118, 129, 138, 148, 157, 164, 166, 183, 184, 189, 197, 217, 224, 227, 231, 239, 241, 257, 265, 277, 281, 285, 287, 289, 294, 314, 316, 333, 334, 344, 346, 372, 375, 378, 381, 403, 413, 426, 435, 488, 493, 529]


In [30]:
df.iloc[[16]]

Observation id  Subs_A  Early win  Subs_B  \
16  1_Poland_Lithuania_Mixed      17          1      16   

                 Behavior Modifier #1 Behavior type     Time Comment  
16  Time_on_offense_teamA         NaN         START  345.633     NaN

In [31]:
# Your list of index points (assumed sorted and valid)
row_index = [1, 9, 16, 25, 29, 37, 40, 47, 59, 65, 72, 81, 87, 93, 96, 103, 105, 107, 118, 129, 138, 148, 157, 164, 166, 183, 184, 189, 197, 217, 224, 227, 231, 239, 241, 257, 265, 277, 281, 285, 287, 289, 294, 314, 316, 333, 334, 344, 346, 372, 375, 378, 381, 403, 413, 426, 435, 488, 493, 529]

# Initialize empty lists to collect DataFrames
teamA_parts = []
teamB_parts = []

# Loop through the index pairs
for i in range(len(row_index) - 1):
    start = row_index[i]
    end = row_index[i + 1]
    
    # Get the slice of df from start to end (excluding end)
    df_slice = df.iloc[start:end]
    
    # Alternate assignment
    if i % 2 == 0:
        teamA_parts.append(df_slice)
    else:
        teamB_parts.append(df_slice)

# Concatenate all slices for each team
df_teamA = pd.concat(teamA_parts).reset_index(drop=True)
df_teamB = pd.concat(teamB_parts).reset_index(drop=True)

In [60]:
teamA_parts

[             Observation id  Subs_A  Early win  Subs_B               Behavior  \
 1  1_Poland_Lithuania_Mixed      17          1      16  Time_on_offense_teamA   
 2  1_Poland_Lithuania_Mixed      17          1      16        Successful_pass   
 3  1_Poland_Lithuania_Mixed      17          1      16        Successful_pass   
 4  1_Poland_Lithuania_Mixed      17          1      16        Successful_pass   
 5  1_Poland_Lithuania_Mixed      17          1      16        Successful_pass   
 6  1_Poland_Lithuania_Mixed      17          1      16        Successful_pass   
 7  1_Poland_Lithuania_Mixed      17          1      16        Forced_turnover   
 8  1_Poland_Lithuania_Mixed      17          1      16                   Pass   
 
   Modifier #1 Behavior type     Time Comment  
 1         NaN         START  295.667     NaN  
 2         Mid         POINT  297.467     NaN  
 3         Mid         POINT  301.267     NaN  
 4         Mid         POINT  306.133     NaN  
 5         Mid      

I think df_teamA (377 rows) is more precise than df_teamA_offense (423 rows).

In [ ]:
df_teamA_offense

In [ ]:
df_teamA

In [ ]:
df_teamB

In [34]:
df['Behavior'].unique()

array(['First half', 'Time_on_offense_teamA', 'Successful_pass',
       'Forced_turnover', 'Pass', 'Score', 'Scoring_attempt',
       'Unforced_turnover', 'Timeout', 'Second half'], dtype=object)

# Analyses
## Formulas for each indicator

### Indicator 2: Attempted scores.
Total scoring attempts.

In [37]:
print(f'Team A had {score_attempts_A} attempted scores. Team B had {score_attempts_B} attempted scores.')

Team A had 24 attempted scores. Team B had 17 attempted scores.


### Indicator 1: Score efficiency.
Ratio of points scored to total scoring attempts.

Score attempts exist of forced/unforced turnovers in the EZ_Off, 'Scoring_attempt', and 'Scores'.

In [ ]:
# Team A total scores and score attempts
scores_A = df_teamA['Behavior'].value_counts().get('Score', 0)
att_scores_A = df_teamA['Behavior'].value_counts().get('Scoring_attempt', 0)
f_turnover_EZ_A = df_teamA[(df_teamA['Behavior'] == 'Forced_turnover') & (df_teamA['Modifier #1'] == 'EZ_Off')].shape[0]
uf_turnover_EZ_A = df_teamA[(df_teamA['Behavior'] == 'Unforced_turnover') & (df_teamA['Modifier #1'] == 'EZ_Off')].shape[0]
score_attempts_A = scores_A + att_scores_A + f_turnover_EZ_A + uf_turnover_EZ_A

# Team B total scores and score attempts
scores_B = df_teamB['Behavior'].value_counts().get('Score', 0)
att_scores_B = df_teamB['Behavior'].value_counts().get('Scoring_attempt', 0)
f_turnover_B = df_teamB[(df_teamB['Behavior'] == 'Forced_turnover') & (df_teamB['Modifier #1'] == 'EZ_Off')].shape[0]
uf_turnover_B = df_teamB[(df_teamB['Behavior'] == 'Unforced_turnover') & (df_teamB['Modifier #1'] == 'EZ_Off')].shape[0]
score_attempts_B = scores_B + att_scores_B + f_turnover_B + uf_turnover_B

In [36]:
score_efficiency_A = scores_A / score_attempts_A
score_efficiency_B = scores_B / score_attempts_B

print(f'The score efficiency of team A is {score_efficiency_A}. The score efficiency of team B is {score_efficiency_B}.')

The score efficiency of team A is 0.625. The score efficiency of team B is 0.8235294117647058.


### Indicator 3: Disc possession.
Ratio of time spent on offense to time spent on defense.

In [43]:
# total_gametime_excl_breaks
# total_a_minutes
# total_b_minutes

print(f'Team A had disc possession for {total_a_minutes / total_gametime_excl_breaks * 100} percent of the time.')
print(f'Team B had disc possession for {total_b_minutes / total_gametime_excl_breaks * 100} percent of the time.')

Team A had disc possession for 53.094501323907885 percent of the time.
Team B had disc possession for 41.63313313463399 percent of the time.


### Indicator 4: Passing accuracy.
Ratio of completed passes to total pass attempts.

In [44]:
# Team A completed passes and total passes
succ_pass_A = df_teamA['Behavior'].value_counts().get('Successful_pass', 0) + scores_A
f_turnover_A = df_teamA['Behavior'].value_counts().get('Forced_turnover', 0)
uf_turnover_A = df_teamA['Behavior'].value_counts().get('Unforced_turnover', 0)
total_pass_A = succ_pass_A + f_turnover_A + uf_turnover_A + att_scores_A

# Team B completed passes and total passes
succ_pass_B = df_teamB['Behavior'].value_counts().get('Successful_pass', 0) + scores_B
f_turnover_B = df_teamB['Behavior'].value_counts().get('Forced_turnover', 0)
uf_turnover_B = df_teamB['Behavior'].value_counts().get('Unforced_turnover', 0)
total_pass_B = succ_pass_B + f_turnover_B + uf_turnover_B + att_scores_B

In [46]:
print(f'Team A had passing accuracy of {succ_pass_A / total_pass_A * 100} percent.')
print(f'Team B had passing accuracy of {succ_pass_B / total_pass_B * 100} percent.')

Team A had passing accuracy of 94.06528189910979 percent.
Team B had passing accuracy of 87.82608695652175 percent.


### Indicator 6: Forced turnovers.
Turnovers caused by active defensive pressure.

`Emma`: Ik kijk naar forced turnovers die ze hebben gemaakt terwijl het andere team op offense was, dus een turnover die ervoor heeft gezorgd dat het betreffende team weer op offense kwam door goede defense.

In [53]:
print(f'Team A caused {f_turnover_B} forced turnovers.')
print(f'Team B caused {f_turnover_A} forced turnovers.')

Team A caused 7 forced turnovers.
Team B caused 10 forced turnovers.


### Indicator 7: Unforced turnovers.
Turnovers due to execution errors or miscommunication.

`Emma`: hier even goed checken hoe ik dit precies bereken, is de forced_turnover van de offense of defense?
`Emma`: Ik kijk naar unforced turnovers die ze zelf hebben gemaakt terwijl ze zelf op offense waren, dus een turnover die ervoor heeft gezorgd dat het betreffende team de aanval verloor door slechte offense.

In [55]:
print(f'Team A caused {uf_turnover_A} unforced turnovers.')
print(f'Team B caused {uf_turnover_B} unforced turnovers.')

Team A caused 7 unforced turnovers.
Team B caused 6 unforced turnovers.


### Indicator 5: Total turnovers.
Total number of times a team lost disc possession.

In [51]:
# Team A total turnovers
total_turnover_A = f_turnover_A + uf_turnover_A

# Team B total turnovers
total_turnover_B = f_turnover_B + uf_turnover_B

In [ ]:
print(f'Team A lost possession {total_turnover_A} times.')
print(f'Team B lost possession {total_turnover_B} times.')

Team A lost possession 17 times.
Team B lost possession 13 times.


### Indicators 8: Number of subs

We know Poland is team A, who had 17 subs.
Lithuania is team B, who had 16 subs.

### Indicator 10: Early wins.
Whether the game finished by maximum points instead of time.

It was an early win (by a few minutes due to reaching 15 points).

### Indicator 9: Possession efficiency.
Number of passes in each successful scoring attempt.

In [77]:
# Step 1: Extract Team A possession intervals (START–STOP)
offense = df[df['Behavior'] == 'Time_on_offense_teamA']
starts = offense[offense['Behavior type'] == 'START']['Time'].tolist()
stops = offense[offense['Behavior type'] == 'STOP']['Time'].tolist()

starts.sort()
stops.sort()

if len(starts) != len(stops):
    raise ValueError("Mismatched START/STOP counts for Team A offense.")

teamA_possessions = list(zip(starts, stops))

# Step 2: Create Team B possessions as gaps between Team A's possessions
teamB_possessions = []
for i in range(min(len(stops), len(starts) - 1)):
    b_start = stops[i]
    b_stop = starts[i + 1]
    if b_stop > b_start:
        teamB_possessions.append((b_start, b_stop))

# Step 3: Get timestamps of all scoring events
score_times = df[df['Behavior'] == 'Score']['Time'].tolist()

# Step 4: Identify possessions that included a score
def get_scoring_possessions(possessions, score_events):
    scoring_possessions = []
    for start, stop in possessions:
        if any((score_time >= start) and (score_time <= stop) for score_time in score_events):
            scoring_possessions.append((start, stop))
    return scoring_possessions

scoring_pos_A = get_scoring_possessions(teamA_possessions, score_times)
scoring_pos_B = get_scoring_possessions(teamB_possessions, score_times)

# Step 5: Count passes in each scoring possession
def count_passes_in_possessions(possessions, team_df):
    pass_counts = []
    for start, stop in possessions:
        passes = team_df[
            (team_df['Time'] >= start) &
            (team_df['Time'] <= stop) &
            (team_df['Behavior'] == 'Successful_pass')
        ]
        pass_counts.append(len(passes))
    return pass_counts

# Step 6: Get pass counts per scoring possession
passes_per_score_A = count_passes_in_possessions(scoring_pos_A, df_teamA)
passes_per_score_B = count_passes_in_possessions(scoring_pos_B, df_teamB)

# Step 7: Compute averages
avg_passes_per_score_A = np.mean(passes_per_score_A) if passes_per_score_A else 0
avg_passes_per_score_B = np.mean(passes_per_score_B) if passes_per_score_B else 0

# Step 8: Display results
print(f"Team A: Avg. passes per scoring possession = {avg_passes_per_score_A:.2f}")
print(f"Team B: Avg. passes per scoring possession = {avg_passes_per_score_B:.2f}")

Team A: Avg. passes per scoring possession = 14.87
Team B: Avg. passes per scoring possession = 3.07


### Indicator 11: Pass rate.
Number of passes made per minute during possession.

Now I want to calculate the number of passes made per minute during possession for each team (A and B). For each team i already have calculated the total passes (total_pass_A and total_pass_B). I also total time on offense (excluding timeouts) for each team (total_a_minutes and total_b_minutes).

In [74]:
pass_rate_A = total_pass_A / total_a_minutes
pass_rate_B = total_pass_B / total_b_minutes

In [75]:
print(f'The pass rate of team A is {pass_rate_A} passes per minute during possession.')
print(f'The pass rate of team B is {pass_rate_B} passes per minute during possession.')

The pass rate of team A is 6.717532094321861 passes per minute during possession.
The pass rate of team B is 2.923399311772485 passes per minute during possession.


### Indicator 12: Possession duration.
Time in seconds a team holds possession before scoring or losing the disc.

So how much time in minutes each offense was for each team, and then the average of that.

In [73]:
# Step 1: Extract Team A possession intervals (START–STOP)
offense = df[df['Behavior'] == 'Time_on_offense_teamA']
starts = offense[offense['Behavior type'] == 'START']['Time'].tolist()
stops = offense[offense['Behavior type'] == 'STOP']['Time'].tolist()

starts.sort()
stops.sort()

if len(starts) != len(stops):
    raise ValueError("Mismatched START/STOP counts for Team A offense.")

teamA_possessions = list(zip(starts, stops))

# Step 2: Create Team B possession intervals (STOP[i] to START[i+1])
teamB_possessions = []
for i in range(min(len(stops), len(starts) - 1)):
    teamB_start = stops[i]
    teamB_stop = starts[i + 1]
    if teamB_stop > teamB_start:
        teamB_possessions.append((teamB_start, teamB_stop))

# Step 3: Extract timeout intervals
timeouts = df[df['Behavior'] == 'Timeout']
to_starts = timeouts[timeouts['Behavior type'] == 'START']['Time'].tolist()
to_stops  = timeouts[timeouts['Behavior type'] == 'STOP']['Time'].tolist()

if len(to_starts) != len(to_stops):
    raise ValueError("Mismatched START/STOP counts for timeouts.")

timeout_intervals = list(zip(to_starts, to_stops))

# Step 4: Function to subtract timeout overlaps from possession intervals
def subtract_timeouts(possessions, timeouts):
    clean_durations = []
    for start, stop in possessions:
        duration = stop - start
        for t_start, t_stop in timeouts:
            overlap_start = max(start, t_start)
            overlap_end = min(stop, t_stop)
            if overlap_start < overlap_end:
                duration -= (overlap_end - overlap_start)
        clean_durations.append(duration)
    return clean_durations

# Step 5: Subtract timeouts from each team's possessions
teamA_durations = subtract_timeouts(teamA_possessions, timeout_intervals)
teamB_durations = subtract_timeouts(teamB_possessions, timeout_intervals)

# Step 6: Calculate averages
avg_teamA_sec = np.mean(teamA_durations)
avg_teamB_sec = np.mean(teamB_durations)

# Step 7: Display results
print(f"Team A average possession duration: {avg_teamA_sec:.2f} seconds ({avg_teamA_sec / 60:.2f} minutes)")
print(f"Team B average possession duration: {avg_teamB_sec:.2f} seconds ({avg_teamB_sec / 60:.2f} minutes)")

Team A average possession duration: 100.33 seconds (1.67 minutes)
Team B average possession duration: 76.88 seconds (1.28 minutes)


In [ ]:
print(f'Team A caused {uf_turnover_A} unforced turnovers.')
print(f'Team B caused {uf_turnover_B} unforced turnovers.')

In [72]:
print(total_a_seconds / len(df_teamA[df_teamA['Behavior type'] == 'START']))

print(total_a_seconds / df_teamA['Behavior type'].value_counts().get('START', 0))

91.21315151515148
91.21315151515148


# Looking at spatial stats

In [83]:
df.loc[df['Behavior'] == 'Score', 'Modifier #1'] = 'EZ_Off'

In [86]:
code_counts = df['Behavior'].value_counts(dropna=False)
code_counts

Behavior
Successful_pass          389
Time_on_offense_teamA     60
Score                     29
Forced_turnover           17
Unforced_turnover         13
Timeout                    8
Pass                       6
Scoring_attempt            4
First half                 2
Second half                2
Name: count, dtype: int64

In [92]:
comments_df = df[df['Comment'].notna()]
comments_df

Observation id  Subs_A  Early win  Subs_B  \
105  1_Poland_Lithuania_Mixed      17          1      16   

                  Behavior Modifier #1 Behavior type    Time  \
105  Time_on_offense_teamA         NaN         START  1530.2   

                                               Comment  
105  was just a too far throw but defensive did tap...

In [84]:
# Count frequency of each spatial location
spatial_counts = df['Modifier #1'].value_counts(dropna=True)

# Display results
print("Spatial Location Summary (from 'Modifier #1'):\n")
print(spatial_counts)

Spatial Location Summary (from 'Modifier #1'):

Modifier #1
Mid       266
RZ_Off     93
RZ_Def     47
EZ_Off     42
EZ_Def      9
Name: count, dtype: int64
